include/cuco/utility/key_generator.cuh

```diff
@@ -94,29 +96,37 @@ struct generate_uniform_fn {
    *
    * @param num Number of elements to generate
    * @param dist Random number distribution
+   * @param seed Random seed
    */
-  __host__ __device__ constexpr generate_uniform_fn(std::size_t num, Dist dist)
-    : num_{num}, dist_{dist}
+  __host__ __device__ constexpr generate_uniform_fn(std::size_t num, Dist dist, std::size_t seed)
+    : num_{num}, dist_{dist}, seed_{seed}
   {
   }
 
   /**
    * @brief Generates a random number of type `T` based on the given `seed`
    *
-   * @param seed Random number generator seed
+   * @param idx Index of the output element
    *
    * @return A resulting random number
    */
-  __host__ __device__ constexpr T operator()(std::size_t seed) const noexcept
+  __host__ __device__ constexpr T operator()(std::size_t idx) const noexcept
   {
     RNG rng;
-    thrust::uniform_int_distribution<T> uniform_dist{1, static_cast<T>(num_ / dist_.value)};
-    rng.seed(seed);
+    // Improved seeding using a linear congruential generator
+    rng.seed(seed_ + idx * 1664525ull + 1013904223ull);
+    // Calculate number of unique keys
+    auto num_unique_keys = cuda::std::max<size_t>(
+      1ull,
+      static_cast<size_t>(
+        cuda::std::ceil(static_cast<double>(num_) / static_cast<double>(dist_.value))));
```

Comment:
we can probably remove https://github.com/NVIDIA/cuCollections/blob/68167400d2c22eab4ac80045eb3bd4121af140de/include/cuco/detail/utility/math.cuh#L38

and use `cuda::std::ceil` instead